src/server/api/contactsApi.js

```diff
@@ -0,0 +1,165 @@
+    return
+  }
+
+  const campaignArray = await r
+    .knex('campaign')
+    .select()
+    .where({ organization_id: orgId, id: campaignId })
+
+  if (campaignArray.length !== 1) {
+    res.writeHead(404)
+    res.end('Not found')
+    return
+  }
+
+  const campaign = campaignArray[0]
+
+  let resp = {}
+
+  switch (req.method) {
+    case 'GET':
+      const count = await r
+        .knex('campaign_contact')
+        .where({ campaign_id: campaignId })
+        .count('*')
+
+      resp = { contacts: { count: parseInt(count[0].count) } }
+      res.writeHead(200, { 'Content-Type': 'application/json' })
+      res.end(JSON.stringify(resp))
+      break
+
+    case 'DELETE':
+      if (campaignaStatusShortCircuit(campaign, res)) {
+        break
+      }
+
+      const deleted = await r
+        .knex('campaign_contact')
+        .delete()
+        .where({ campaign_id: campaignId })
+
+      resp = { contacts: { deleted } }
+      res.writeHead(200, { 'Content-Type': 'application/json' })
+      res.end(JSON.stringify(resp))
+      break
+
+    case 'POST':
+      if (campaignaStatusShortCircuit(campaign, res)) {
+        break
+      }
+
+      const { validationStats, validatedData } = getValidatedData(req.body, [])
+
+      const successResponse = {
+        invalid: _.concat(
+          validationStats.invalidCellRows,
+          validationStats.missingCellRows
+        ),
+        dupesInBatch: validationStats.dupeCount,
+        numberSubmitted: req.body.length
+      }
+
+      const standardFields = [
+        'first_name',
+        'last_name',
+        'cell',
+        'external_id',
+        'zip'
+      ]
+
+      const contactsToSavePromises = validatedData.map(async contact => {
+        const contactToSave = _.pick(contact, standardFields)
+        const customFields = _.omit(contact, standardFields)
+        contactToSave.custom_fields = JSON.stringify(customFields)
+        contactToSave.timezone_offset = await getTimezoneByZip(
+          contactToSave.zip
+        )
+        return contactToSave
+      })
+
+      const contactsToSave = await Promise.all(contactsToSavePromises)
+
+      await r.knex
+        .transaction(async tr => {
+          const existingContacts = await tr('campaign_contact')
+            .select('cell')
+            .whereIn('cell', contactsToSave.map(contact => contact.cell))
+            .andWhere({ campaign_id: campaignId })
+
+          const existingContactCells = new Set(
+            existingContacts.map(contact => contact.cell)
+          )
+
+          const dedupedContactsToSave = _.filter(
+            contactsToSave,
+            contact => !existingContactCells.has(contact.cell)
+          )
+
+          successResponse['dupesInCampaign'] =
+            contactsToSave.length - dedupedContactsToSave.length
+
+          await tr
+            .batchInsert(
+              'campaign_contact',
+              dedupedContactsToSave.map(row => {
+                row.campaign_id = campaignId
```

Comment:
wouldn't it be better to do this above when you're originally creating the contact?,  then you don't have to do a map at all, right?

Reply:
👍